src/basic/process-util.c

```diff
@@ -698,6 +698,62 @@ int wait_for_terminate_and_warn(const char *name, pid_t pid, bool check_exit_cod
         return -EPROTO;
 }
 
+int wait_for_terminate_with_timeout(pid_t pid, usec_t timeout) {
+        sigset_t mask;
+        int r;
+        siginfo_t status;
+        usec_t until;
+
+        assert_se(sigemptyset(&mask) == 0);
+        assert_se(sigaddset(&mask, SIGCHLD) == 0);
+
+        /* Drop into a sigtimewait-based timeout. Waiting for the
+         * remount pid to exit. */
+        until = now(CLOCK_MONOTONIC) + timeout;
+        for (;;) {
+                usec_t n;
+                struct timespec ts;
+
+                n = now(CLOCK_MONOTONIC);
+                if (n >= until)
+                        break;
+
+                zero(status);
+
+                r = sigtimedwait(&mask, NULL, timespec_store(&ts, until - n));
+                if (r < 0) {
+                        if (errno == EAGAIN) {
+                                /* Timed out, remount child is likely hung,
+                                 * but we can try. */
+                                if (waitid(P_PID, pid, &status, WEXITED|WNOHANG) == 0) {
+                                        if (status.si_pid == pid)
+                                                /* This is the correct child */
+                                                return (status.si_status == EXIT_SUCCESS) ? 0 : -EPROTO;
+                                        else
+                                                /* Fail, we should return the timeout */
+                                                break;
+                                }
+                        } else if (errno == EINTR)
+                                /* Received a different signal - Retry */
+                                continue;
+                } else {
+                        /* Received a child signal, checking if it is the one
+                         * we are watching */
+                       if (waitid(P_PID, pid, &status, WEXITED|WNOHANG) == 0) {
+                               if (status.si_pid == pid)
+                                       /* This is the correct child */
+                                       return (status.si_status == EXIT_SUCCESS) ? 0 : -EPROTO;
```

Comment:
so, ultimately both branches where we call waitid() are pretty much the same no? I think we can merge them, i.e. let's just wait with sigtimedwait(), and when it returns, before checking its result check with waitid() if our child died. And if so, let's just propagate that, regardless if sigtimedwait() failed or not. Afterwards, lets check the original error sigtimedwait() returned, and if it's anything but EAGAIN/EINTR let's propagate it, otherwise try again